### DISCUSSION TITLE
V8 memory leak?

### DISCUSSION THREAD

**Discussion by guest271314** [denoland/deno#17236] [Q&A] [answered]
When I pass input array to a Deno Native Messaging host having length `4096*4` which when stringified as JSON has length `81921` Deno Native Messaging host memory usage grows to 1GB then the script exits. The same thing happens with Node.js Native Messaging host. Both are dependent on V8 JavaScript engine.

Not so long ago I was able to pass `new Array(200000)` to each Native Messaging host and get echoed stdout.

QuickJS, C, and C++ Native Messaging hosts can each process 200000 array input. Python exists when input is 175000.

Is there a recent change to V8 which limits input to stadin and/or stdout; caches or buffers stdin or stdout?

**Comment by guest271314** [comment]
@bartlomieju @ayame113 @lucacasonato @crowlKats @ry 

I'm pretty sure this is a bug in V8. 

Unless one of you can explain why V8 is writing `1567386741` to the `uint32` when the actual length of the input is `65536`?

When input to stdin is equal to or greater than 65536 memory usage exponentially increases to upwards of 700MB-1GB - this happens only when using Node.js or Deno JavaScript engines . 

Using the same code with QuickJS JavaScript engine, C and C++ Native Messaging hosts I get no such unexpected behaviour.

I narrowed down the range when this happens and write what happens to a file.

```
port.postMessage(new Array((4096*3)+818));
```
```
65531
```
is written to log.txt, which is expected

```
JSON.stringify(new Array((4096*3)+818)).length // 65531
```

when I add 1 to that input array
```
port.postMessage(new Array((4096*3)+819));
```
```
1567386741
```
is written to log.txt.

The relevant Deno code
 
```
async function getMessage() {
  const header = new Uint32Array(1);
  await Deno.stdin.read(header);
  await Deno.writeTextFile("log.txt", header[0]);
  const output = new Uint8Array(header[0]);
  await Deno.stdin.read(output);
  return output;
}
```

Node.js code
``` 
function getMessage() {
   const header = new Uint32Array(1);
  readSync(0, header);
  const content = new Uint8Array(header[0]);
  writeFileSync('log.txt', JSON.stringify(header[0]));
  readSync(0, content);
  return content;
}
```

**Comment by 0f-0b** [comment, marked as answer]
It fails when receiving the second message because `Deno.stdin.read(buffer)` is [not guaranteed to read exactly `buffer.byteLength` bytes](https://deno.land/api@v1.29.1?s=Deno.Reader), and thus the first message was not fully read. `BufReader` from std provides a convenient way to [ensure that the whole buffer is filled](https://deno.land/std@0.170.0/io/buf_reader.ts?s=BufReader#method_readFull_0).

```js
import { BufReader } from "https://deno.land/std@0.170.0/io/buf_reader.ts";
import { BufWriter } from "https://deno.land/std@0.170.0/io/buf_writer.ts";

const stdin = new BufReader(Deno.stdin);
const stdout = new BufWriter(Deno.stdout);

async function getMessage() {
  const header = new Uint32Array(1);
  await stdin.readFull(new Uint8Array(header.buffer));
  const message = new Uint8Array(header[0]);
  await stdin.readFull(message);
  return message;
}

async function sendMessage(message) {
  const header = new Uint32Array([message.length]);
  await stdout.write(new Uint8Array(header.buffer));
  await stdout.write(message);
  await stdout.flush();
}
```

I am not aware of any built-in equivalent in Node, but the following should work.

```js
function readFullSync(fd, buf) {
  let offset = 0;
  while (offset < buf.byteLength) {
    offset += readSync(fd, buf, { offset });
  }
  return buf;
}

function getMessage() {
  const header = new Uint32Array(1);
  readFullSync(0, header);
  const content = new Uint8Array(header[0]);
  readFullSync(0, content);
  return content;
}
```

<details>
<summary>Original answer</summary>

I cannot reproduce this locally.

I installed the [native messaging example](https://github.com/GoogleChrome/chrome-extensions-samples/tree/main/mv2-archive/api/nativeMessaging) and modified it as follows. When I visit `main.html`, `65536` is written to `log.txt` as expected.

```js
// app/main.js
const port = chrome.runtime.connectNative("com.google.chrome.example.echo");
port.postMessage(new Array((4096 * 3) + 819));

// host/native-messaging-example-host
#!/usr/bin/env -S deno run --allow-write=log.txt

async function getMessage() {
  const header = new Uint32Array(1);
  await Deno.stdin.read(header);
  await Deno.writeTextFile("log.txt", header[0]);
  const output = new Uint8Array(header[0]);
  await Deno.stdin.read(output);
  return output;
}

await getMessage();
```

</details>

**Reply by guest271314** [reply]
Can you include the complete extension you tested here, or create a gist with the complete code?

Not sure why I can consistently get the incorrect result in the code I tested?

- https://github.com/guest271314/native-messaging-deno
- https://github.com/guest271314/native-messaging-nodejs

**Reply by guest271314** [reply]
How did you even install the Native Messaging host using MV2 extension?

**Reply by 0f-0b** [reply]
> Can you include the complete extension you tested here, or create a gist with the complete code?

[Sure](https://gist.github.com/0f-0b/97ad6c7e6380d8142cea1351a619a4c9). This is the first time I have heard about native messaging though, so please point out if I am doing something wrong.

> How did you even install the Native Messaging host using MV2 extension?

By following the instructions in the [readme file included in the example](https://github.com/GoogleChrome/chrome-extensions-samples/blob/main/mv2-archive/api/nativeMessaging/README.txt).

**Reply by guest271314** [reply]
MV2 extensions are deprecated. 

I managed to cobble together what you did locally.

Did you notice that the Native Messaging host exits immediately? 

Try the above links.

**Reply by guest271314** [reply]
In fact you should be able to just remove your `await getMessage()` and replace that with the following and observe the `console` at the HTML page, where the Native Messaging host exists immediately. Though I suggest installing the repositories I linked above.

Since you are using `chrome.runtime.connectNative()` the host should not exit immediately.

```
async function sendMessage(message) {
  const header = Uint32Array.from(
    {
      length: 4,
    },
    (_, index) => (message.length >> (index * 8)) & 0xff
  );
  const output = new Uint8Array(header.length + message.length);
  output.set(header, 0);
  output.set(message, 4);
  await Deno.stdout.write(output.buffer);
}

async function main() {
  while (true) {
    const message = await getMessage();
    await sendMessage(message);
    gc();
  }
}

try {
  main();
} catch (e) {
  Deno.exit();
}
```

**Reply by guest271314** [reply]
Now, using the existing code you have, with the above substituted for your original `await getMessage()` you'll notice that `deno` exits after writing to the text file and `sendMessage` is never reached.

Compare that result to the same algorithm using non-V8 based JavaScript engine, e.g., QuickJS https://github.com/guest271314/native-messaging-quickjs, and C https://github.com/guest271314/native-messaging-cpp, and C++ https://github.com/guest271314/native-messaging-cpp versions. The only ones that exit are the V8 based JavaScript engines.

**Reply by 0f-0b** [reply]
I see. It fails when receiving the second message. This is because `Deno.stdin.read(buffer)` is [not guaranteed to read exactly `buffer.byteLength` bytes](https://deno.land/api@v1.29.1?s=Deno.Reader), and thus the first message was not fully read. `BufReader` from std provides a convenient way to [ensure that the whole buffer is filled](https://deno.land/std@0.170.0/io/buf_reader.ts?s=BufReader#method_readFull_0).

```js
import { BufReader } from "https://deno.land/std@0.170.0/io/buf_reader.ts";
import { BufWriter } from "https://deno.land/std@0.170.0/io/buf_writer.ts";

const stdin = new BufReader(Deno.stdin);
const stdout = new BufWriter(Deno.stdout);

async function getMessage() {
  const header = new Uint32Array(1);
  await stdin.readFull(new Uint8Array(header.buffer));
  const message = new Uint8Array(header[0]);
  await stdin.readFull(message);
  return message;
}

async function sendMessage(message) {
  const header = new Uint32Array([message.length]);
  await stdout.write(new Uint8Array(header.buffer));
  await stdout.write(message);
  await stdout.flush();
}
```

**Reply by guest271314** [reply]
Include `onMessage` and `onDisconnect` handlers in your existing code to observe V8-based JavaScript engines exit when input length to stdin is equal to or greater than 65536

```
const port = chrome.runtime.connectNative("com.google.chrome.example.echo");
port.onMessage.addListener((message) => console.log(message));
port.onDisconnect.addListener((p) => console.log(p));
port.postMessage(new Array(4096 * 4));
```

**Reply by guest271314** [reply]
> I see. It fails when receiving the second message. This is because `Deno.stdin.read(buffer)` is [not guaranteed to read exactly `buffer.byteLength` bytes](https://deno.land/api@v1.29.1?s=Deno.Reader), and thus the first message was not fully read. `BufReader` from std provides a convenient way to [ensure that the whole buffer is filled](https://deno.land/std@0.170.0/io/buf_reader.ts?s=BufReader#method_readFull_0).

Ahh. That changes everything. Thanks. Now able to write and read `new Array(200000)`.

What is the equivalent in the Node.js version?

In pertinent part

```


#!/usr/bin/env -S ./node --max-old-space-size=6 --jitless --expose-gc --v8-pool-size=1
// Node.js Native Messaging host
// guest271314, 10-9-2022
import {readSync} from 'node:fs';
// Node.js Native Messaging host constantly increases RSS during usage
// https://github.com/nodejs/node/issues/43654
process.env.UV_THREADPOOL_SIZE = 1;
// Process greater than 65535 length input
// https://github.com/nodejs/node/issues/6456
// https://github.com/nodejs/node/issues/11568#issuecomment-282765300
process.stdout._handle.setBlocking(true);

function getMessage() {
  const header = new Uint32Array(1);
  readSync(0, header);
  const content = new Uint8Array(header[0]);
  readSync(0, content);
  return content;
}
```

**Reply by guest271314** [reply]
@0f-0b How do I mark your answer as answered?

**Reply by 0f-0b** [reply]
> What is the equivalent in the Node.js version?

I am too unfamiliar with Node to know if there is a built-in equivalent, but the following might work.

```js
function readFullSync(fd, buf) {
  let offset = 0;
  while (offset < buf.byteLength) {
    offset += readSync(fd, buf, { offset });
  }
  return buf;
}

function getMessage() {
  const header = new Uint32Array(1);
  readFullSync(0, header);
  const content = new Uint8Array(header[0]);
  readFullSync(0, content);
  return content;
}
```

> How do I mark your answer as answered?

I edited my answer to reflect the actual solution. Or perhaps you could post your finished code in a new answer and mark that as answered?

**Reply by guest271314** [reply]
That does work. Thanks again.

**Reply by guest271314** [reply]
Can you kindly also include that code in your accepted answer as well re my subsequent Node.js equivalent question?
